src/perl/lib/wtsi_clarity/clarity/process.pm

```diff
@@ -16,6 +21,126 @@ Readonly::Scalar my $OUTPUT_LIMSID       => q{./output[@output-type!="ResultFile
 Readonly::Scalar my $ALL_CONTAINERS      => q{art:details/art:artifact/location/container/@uri};
 ## use critic
 
+has '_parent' => (
+  is => 'ro',
+  isa => 'Object',
+  required => 1,
+  init_arg => 'parent',
+);
+
+has '_config' => (
+  is => 'ro',
+  isa => 'wtsi_clarity::util::config',
+  required => 0,
+  init_arg => undef,
+  lazy_build => 1,
+);
+sub _build__config {
+  my $self = shift;
+  return $self->_parent->config;
+}
+
+has '_request' => (
+  is => 'ro',
+  isa => 'wtsi_clarity::util::request',
+  required => 0,
+  init_arg => undef,
+  lazy_build => 1,
+);
+sub _build__request {
+  my $self = shift;
+  return $self->_parent->request;
+}
+
+has 'xml' => (
+  is => 'rw',
+  isa => 'XML::LibXML::Document',
+  required => 1,
+  init_arg => 'xml',
+  handles => {
+    find               => 'find',
+    findvalue          => 'findvalue',
+    findnodes          => 'findnodes',
+    toString           => 'toString',
+    getDocumentElement => 'getDocumentElement',
+    createElementNS    => 'createElementNS',
+  },
+);
+
+has 'input_artifacts' => (
+  isa             => 'XML::LibXML::Document',
+  is              => 'rw',
+  required        => 0,
+  lazy_build      => 1,
+);
+sub _build_input_artifacts {
+  my $self = shift;
+
+  my $input_node_list = $self->findnodes($INPUT_ARTIFACT_URIS_PATH);
+  my @input_uris = uniq(map { $_->getValue } $input_node_list->get_nodelist);
+
+  return $self->_request->batch_retrieve('artifacts', \@input_uris);
+}
+
+sub find_parent {
+  my ($self, $needle_process_name, $child_process_url) = @_;
+
+  my $parent_processes = $self->_find_parent($needle_process_name, $child_process_url);
+
+  return $parent_processes;
+}
+
+sub _find_parent {
+  my ($self, $needle_process_name, $process_url, $found_processes) = @_;
+
+  if (!defined $found_processes) {
+    $found_processes = {};
+  }
+
+  my $current_process = $self->_parent->fetch_and_parse($process_url);
+
+  my $current_process_name = $current_process->findvalue($PROCESS_TYPE);
+
+  if ($current_process_name eq $needle_process_name) {
+    $found_processes->{$process_url} = q{};
+  } else {
+    my $parent_uris = $current_process->findnodes($PARENT_PROCESS_PATH);
+
+    if ($parent_uris->size() > 0) {
+      my @uniq_uris = map { $_->getValue() } $parent_uris->get_nodelist();
+
+      foreach my $uri (@uniq_uris) {
+        $self->_find_parent($needle_process_name, $uri, $found_processes);
+      }
+    }
+  }
+
+  my @found_processes = keys %{$found_processes};
+
+  return \@found_processes;
+}
+
+sub find_previous_process {
```

Comment:
It's finding a previous process of an artifact ($artifact_limsid) that has a particular process_name ($process_name)

So it's like `find_previous_process_of_artifactlimsid_by_name`